### PR TITLE
[Navigation] Refactoring of CarDirectionsEngine and PedestrianDirectionsEngine

### PR DESCRIPTION
**Pull request by AntonM030481**
Reason: 
A lot of duplicated code which is not easy to read.
Results are not easy to understand.

1. Preparation by updating pedestrian turn tests.
2. Make functions the same, but hold different names. Line by line comparison can be seen here.
3. Remove MakeTurnAnnotationPedestrian.
4. Added comments for input and output data structures.

Logical continuation is changing of CarDirectionsEngine and PedestrianDirectionsEngine (childs of DirectionsEngine).
Currently the only method they have - almost identical Generate().

Suggested changes:
MakeTurnAnnotation() - becomes method of DirectionsEngine.
All VehicleType dependent logic is incapsulated in virtual methods GetTurnDirection(), FixupTurns().
Existing functions become methods of CarDirectionsEngine and PedestrianDirectionsEngine.

If this next step is OK, I will make changes in this PR.

### CODE REVIEW

**Comment by vng** [comment]
@AntonM030481 You are doing great, so do not hesitate to refactor according to your vision of a better code.

**Comment by AntonM030481** [comment]
No further changes are planned. Ready for review.

**Comment by vng** [comment]
Is it necessary to update icu and boost here?

File: routing/directions_engine.cpp [hunk 1]
```diff
@@ -223,4 +224,182 @@ void DirectionsEngine::FillPathSegmentsAndAdjacentEdgesMap(
...
+    // Street names contain empty names too for avoiding of freezing of old street name while
+    // moving along unnamed street.
+    streets.emplace_back(max(junctions.size(), static_cast<size_t>(1)) - 1, loadedSegmentIt->m_name);
+
+    // Turns information.
+    if (!junctions.empty() && skipTurnSegments == 0)
+    {
+      auto const outgoingSegmentDist = distance(loadedSegments.begin(), loadedSegmentIt);
+      CHECK_GREATER(outgoingSegmentDist, 0, ());
+      auto const outgoingSegmentIndex = static_cast<size_t>(outgoingSegmentDist);
```
Comment (vng) [inline routing/directions_engine.cpp:358]: Can assign directly without intermediate outgoingSegmentDist here

File: routing/turns_generator.cpp [hunk 2]
```diff
@@ -828,25 +268,53 @@ double CalcTurnAngle(IRoutingResult const & result,
...
     }
+    else if (nodes.isCandidatesAngleValid)
+      ASSERT(false, ("Can't match any candidate with firstOutgoingSegment but isCandidatesAngleValid == true."));
+    else
+    {
+      LOG(LWARNING, ("Can't match any candidate with firstOutgoingSegment and isCandidatesAngleValid == false"));
+      if (candidates.size() == 1)
+      {
+        ASSERT(candidates.front().m_segment.GetMwmId() != firstOutgoingSeg.GetMwmId(), ());
+        ASSERT(candidates.front().m_segment.GetSegmentIdx() == firstOutgoingSeg.GetSegmentIdx() && candidates.front().m_segment.IsForward() == firstOutgoingSeg.IsForward(), ());
```
Comment (vng) [inline routing/turns_generator.cpp:297]: Split on 2 lines

File: routing/turns_generator.cpp [hunk 3]
```diff
@@ -828,25 +268,53 @@ double CalcTurnAngle(IRoutingResult const & result,
...
   auto IsFirstOutgoingSeg = [&firstOutgoingSeg](TurnCandidate const & turnCandidate) { return turnCandidate.m_segment == firstOutgoingSeg; };
-  if (find_if(candidates.begin(), candidates.end(), IsFirstOutgoingSeg) == candidates.end())
+  auto & candidates = nodes.candidates;
+  auto it = find_if(candidates.begin(), candidates.end(), IsFirstOutgoingSeg);
+  if (it == candidates.end())
   {
-    double const turnAngle = CalcOneSegmentTurnAngle(turnInfo);
-    auto DoesAngleMatch = [&turnAngle](TurnCandidate const & turnCandidate) { return base::AlmostEqualAbs(turnCandidate.m_angle, turnAngle, 0.001); };
+    auto DoesAngleMatch = [&turnAngle](TurnCandidate const & turnCandidate)
+    { return base::AlmostEqualAbs(turnCandidate.m_angle, turnAngle, 0.001) || abs(turnCandidate.m_angle) + abs(turnAngle) > 359.999; };
```
Comment (vng) [inline routing/turns_generator.cpp:281]: Split like normal block:
{
...
};

**Comment by AntonM030481** [comment]
Reset + push --force to remove accidental icu and boost changes.

File: routing/car_directions.cpp [hunk 4]
```diff
@@ -1,81 +1,630 @@
...
+    LOG(LDEBUG, (GetTurnString(t.m_turn), ":", t.m_index, t.m_sourceName, "-",
+                 t.m_targetName, "exit:", t.m_exitNum));
+#endif
+}
 
-  if (routeEdges.empty())
-    return false;
+void FixupCarTurns(std::vector<geometry::PointWithAltitude> const & junctions, Route::TTurns & turnsDir)
+{
+  uint32_t turn_index = static_cast<uint32_t>(junctions.size() - 1);
```
[the remark this answers is not in the document]
Comment (AntonM030481) [reply routing/car_directions.cpp:33]: Yes. Added assert just in case.

File: routing/turns_generator.cpp (hunk 2 above)
Comment (AntonM030481) [reply routing/turns_generator.cpp:297]: ok

File: routing/turns_generator.cpp (hunk 3 above)
Comment (AntonM030481) [reply routing/turns_generator.cpp:281]: ok

File: routing/directions_engine.cpp (hunk 1 above)
Comment (AntonM030481) [reply routing/directions_engine.cpp:358]: ok

File: routing/car_directions.cpp [hunk 5]
```diff
@@ -1,81 +1,630 @@
...
+ */
+bool KeepRoundaboutTurnByHighwayClass(TurnCandidates const & possibleTurns,
+                                      TurnInfo const & turnInfo, NumMwmIds const & numMwmIds)
+{
+  Segment firstOutgoingSegment;
+  bool const validFirstOutgoingSeg = turnInfo.m_outgoing->m_segmentRange.GetFirstSegment(numMwmIds, firstOutgoingSegment);
+
+  for (auto const & t : possibleTurns.candidates)
+  {
+    if (!validFirstOutgoingSeg || t.m_segment == firstOutgoingSegment)
```
[the remark this answers is not in the document]
Comment (AntonM030481) [reply routing/car_directions.cpp:178]: In fact this check can be removed.
It was checked before in GetTurnDirectionBasic().

**Review by vng — APPROVED** [review]

**Comment by AntonM030481** [comment]
Ready for merge.

**Comment by vng** [comment]
Please, rebase and resolve conflicts.

**Comment by AntonM030481** [comment]
I am not an expert in git.
This branch is 11 commits ahead of organicmaps/organicmaps:master.
Does it mean that there are no conflicts now?

**Comment by vng** [comment]
@AntonM030481 Don't bother, seems like I can squash and merge. You have got conflicts between your commits some how ..

**Comment by AntonM030481** [comment]
This is commit message for squash:
```
[Routing] Refactoring of CarDirectionsEngine and PedestrianDirectionsEngine

Almost identical Generate()
from CarDirectionsEngine and PedestrianDirectionsEngine
moved to parent class.

MakeTurnAnnotation() - became method of DirectionsEngine.

All VehicleType dependent logic was incapsulated in virtual methods:
- GetTurnDirection()
- FixupTurns().

According existing functions became methods of
CarDirectionsEngine and PedestrianDirectionsEngine.

VehicleType dependent functions moved
from turns_generator to according DirectionsEngines.

Pedestrian turn tests updated.
```

**Comment by AntonM030481** [comment]
@vng Thank you!

**Pull request merged**